parlai/agents/bert_qa/bert_qa.py

```diff
@@ -0,0 +1,443 @@
+            for end_index in end_indexes:
+                # We could hypothetically create invalid predictions
+                if end_index < start_index:
+                    continue
+                if start_context_position and start_index < start_context_position:
+                    # answer in the question
+                    continue
+                length = end_index - start_index + 1
+                if length > self.text_truncate:
+                    continue
+                prelim_predictions.append(
+                    _PrelimPrediction(
+                        start_index=start_index,
+                        end_index=end_index,
+                        start_logit=start_logits[start_index],
+                        end_logit=end_logits[end_index],
+                    )
+                )
+
+        prelim_predictions = sorted(
+            prelim_predictions,
+            key=lambda x: (x.start_logit + x.end_logit),
+            reverse=True,
+        )
+
+        _NbestPrediction = collections.namedtuple(  # pylint: disable=invalid-name
+            "NbestPrediction", ["text", "start_logit", "end_logit"]
+        )
+
+        nbest = []
+        for pred in prelim_predictions:
+            if len(nbest) >= self.opt["n_best_size"]:
+                break
+
+            if pred.start_index > 0:  # this is a non-null prediction
+                answer_tokens = self.dict.vec2txt(
+                    input_ids[pred.start_index:pred.end_index]
+                )
+                answer_text = self._clean_WordPieces(answer_tokens)
+                nbest.append(
+                    _NbestPrediction(
+                        text=answer_text,
+                        start_logit=pred.start_logit,
+                        end_logit=pred.end_logit,
+                    )
+                )
+
+        if len(nbest) > 0:
+            # return the best prediciton
+            return nbest[0].text
+        else:
+            # return an empty prediction
+            return ""
+
+    def report(self):
+        base = super().report()
+        m = {}
+        examples = self.metrics["examples"]
+        if examples > 0:
+            m["examples"] = examples
+            m["loss"] = self.metrics["loss"]
+            m["mean_loss"] = self.metrics["loss"] / examples
+
+        for k, v in m.items():
+            # clean up: rounds to sigfigs and converts tensors to floats
+            base[k] = round_sigfigs(v, 4)
+        return base
+
+    def train_step(self, batch):
+
+        tensors = self._vectorize(batch["observations"])
+        loss = self.model(*tensors)
+
+        if self.n_gpu > 1:
+            loss = loss.mean()  # mean() to average on multi-gpu.
+
+        self.metrics["examples"] += len(batch["observations"])
+        self.metrics["loss"] += loss
+
+        self.backward(loss)
+        self.update_params()
+        self.zero_grad()
+
+        # predictions
+        with torch.no_grad():
```

Comment:
this doesn't make much sense to me; why wouldn't we want to backpropagate through the model here?

Reply:
rather, should probably collect the b_start_logits and b_end_logits during the first forward pass above rather than doing two forward passes during the train step, one for learning and one for predictions

Reply:
unfortunately the forward function returns the loss during training and the logits during evaluation. This is the reason for the double call